maven-slf4j-provider/src/main/java/org/slf4j/impl/MavenLoggerFactory.java

```diff
@@ -0,0 +1,82 @@
+package org.slf4j.impl;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import org.apache.maven.logwrapper.LogLevelRecorder;
+import org.apache.maven.logwrapper.MavenSlf4jWrapperFactory;
+import org.slf4j.Logger;
+
+import java.util.Optional;
+
+/**
+ * LogFactory for Maven which can create a simple logger or a one which, if set, fails the build on a threshold.
+ */
+public class MavenLoggerFactory extends SimpleLoggerFactory implements MavenSlf4jWrapperFactory
+{
+    private LogLevelRecorder logLevelRecorder = null;
+
+    @Override
+    public void setLogLevelRecorder( LogLevelRecorder logLevelRecorder )
+    {
+        if ( this.logLevelRecorder != null )
+        {
+            throw new IllegalStateException( "Maven logger fail level has already been set." );
+        }
+
+        this.logLevelRecorder = logLevelRecorder;
+    }
+
+    @Override
+    public Optional<LogLevelRecorder> getLogLevelRecorder()
+    {
+        return Optional.ofNullable( logLevelRecorder );
+    }
+
+    /**
+     * Return an appropriate {@link MavenSimpleLogger} instance by name.
+     */
+    @Override
+    public Logger getLogger( String name )
+    {
+        Logger simpleLogger = loggerMap.get( name );
+        if ( simpleLogger != null )
+        {
+            return simpleLogger;
+        }
+        else
+        {
+            Logger newInstance = getNewLoggingInstance( name );
+            Logger oldInstance = loggerMap.putIfAbsent( name, newInstance );
+            return oldInstance == null ? newInstance : oldInstance;
+        }
+    }
+
+    private Logger getNewLoggingInstance( String name )
```

Comment:
I need to dive into this. It is likely this method will be called a lot, and every time you check if logLeverRecorder is null, even though this won't change. 
Java 8 MethodReference is hopefully the answer to this.

Reply:
I think it's called a lot as well, once for each class that requests a logger is my guess.
How do you propose to utilize a MethodReference here?

Reply:
A null check isn't that expensive, is it?

But if you fear it becomes a bottleneck, you might want to have an instance variable of type `Supplier<Logger>`. By default, it would be `() -> return new MavenSimpleLogger( name )`, but inside `setLogLevelRecorder`, you could populate it with `() -> return new MavenFailLevelLogger( name, logLevelRecorder )`.

Reply:
it is a repeating null-check and the result is always the same, just looking for a way to optimize it (if possible)

Reply:
I ran a small performance test with 3 scenarios. The (current) null-check, a Function/Supplier like Maarten suggested and the removal of any check (the old situation, without any severity logging features). All of them are very quick and the differences are negligible, I guess JIT will also help a lot here.

Personally I think we should go for readability/simplicty, so I would keep the null-check. Even though the Function one is quite elegant.

Reply:
Cool,I that case I agree that readability is more important